src/utilities/substituteFragmentArguments.ts

```diff
@@ -0,0 +1,79 @@
+import type { Maybe } from '../jsutils/Maybe.js';
+
+import type {
+  ArgumentNode,
+  FragmentDefinitionNode,
+  FragmentSpreadNode,
+  SelectionSetNode,
+  ValueNode,
+  VariableDefinitionNode,
+} from '../language/ast.js';
+import { Kind } from '../language/kinds.js';
+import { visit } from '../language/visitor.js';
+
+// TODO: follow up on https://github.com/graphql/graphql-js/pull/3835/files#r1101010604
+
+/**
+ * Replaces all fragment argument values with non-fragment-scoped values.
+ *
+ * NOTE: fragment arguments are scoped to the fragment they're defined on.
+ * Therefore, after we apply the passed-in arguments, all remaining variables
+ * must be either operation defined variables or explicitly unset.
+ */
+export function substituteFragmentArguments(
+  def: FragmentDefinitionNode,
+  fragmentSpread: FragmentSpreadNode,
+): SelectionSetNode {
+  const argumentDefinitions = def.variableDefinitions;
+  if (argumentDefinitions == null || argumentDefinitions.length === 0) {
+    return def.selectionSet;
+  }
+  const argumentValues = fragmentArgumentSubstitutions(
+    argumentDefinitions,
+    fragmentSpread.arguments,
+  );
+  return visit(def.selectionSet, {
+    Variable(node) {
+      return argumentValues.get(node.name.value);
+    },
+  });
+}
+
+export function fragmentArgumentSubstitutions(
+  variableDefinitions: ReadonlyArray<VariableDefinitionNode>,
+  argumentValues: Maybe<ReadonlyArray<ArgumentNode>>,
+): Map<string, ValueNode> {
+  const substitutions = new Map<string, ValueNode>();
+  if (argumentValues) {
+    for (const argument of argumentValues) {
+      substitutions.set(argument.name.value, argument.value);
+    }
+  }
+
+  for (const variableDefinition of variableDefinitions) {
+    const argumentName = variableDefinition.variable.name.value;
+    if (substitutions.has(argumentName)) {
+      continue;
+    }
+
+    const defaultValue = variableDefinition.defaultValue;
+    if (defaultValue) {
+      substitutions.set(argumentName, defaultValue);
+    } else {
+      // We need a way to allow unset arguments without accidentally
+      // replacing an unset fragment argument with an operation
+      // variable value. Fragment arguments must always have LOCAL scope.
+      //
+      // To remove this hack, we need to either:
```

Comment:
The issue here presents itself in that fragments can define arguments like

```
fragment x($id: ID!) on Y { todo(id: $id) { text } }
```

Which can be passed in through fragment spreads

```
...x(id: '2')
```

now what happens when we spread `x` and there is an `$id` present on the document itself, if we aren't using `__UNSET` it would leverage that one which we don't want. There might be a better solution in carrying `variables` around as we visit the `FragmentSpread` and the accompanying `FragmentDefinition`

Reply:
An alternative I have been playing around with is to basically create a new `argumentsContext` when we go from a FragmentSpread into Fields within the collectFields, we would treat the `FragmentDefinition` variables similar to an operation where we calculate a scope and pass it into the `FieldInfo` so we can sue that in execute. I do think that the over-arching refactor could be quite big though.

Reply:
Another attempt https://github.com/JoviDeCroock/graphql-js/compare/fragment-args-execution-2024...JoviDeCroock:graphql-js:collect-fields-alternative?expand=1

Reply:
Relevant https://benjie.dev/rfcs/1056

Reply:
> We need to introduce a concept of shadowing:
> - when a fragment defines a variable that is in the parent scope but not given
>    in the fragment-spread we need to look at this variable as undefined and check
>    whether the definition has a defaultValue, if not remove it from the variableValues.
> - when a fragment does not define a variable we need to copy it over from the parent
>    scope as that variable can still get used in spreads later on in the selectionSet.
> - when a value is passed in through the fragment-spread we need to copy over the key-value
>    into our variable-values.

I'm not sure this is *quite* correct. Specifically, "remove it from the variableValues": are you saying "remove it from the variableValues *to use for this specific fragment*" or "remove from variableValues for all future execution"? I can't quite tell yet.

Those two imply very different semantic "scopes". What we want to avoid is adding in some form of "contextual" (lexical?) scope. Operation variables currently have "dynamic" scope, i.e. the definition of the variable (including its type and default value) depends on *which* operation you're executing. We absolutely don't want a *second* form of dynamic scoping, where child fragment spreads "inherit" their parent fragments' argument definitions. Nor do we want a parent fragment to "wipe out" an operations' variable definition, *except* within the context of *that one fragment*.

Reply:
It removes it from the new scope i.e. `variableValues` is altered when we are collecting field for said fragment-spread and subsequent nested fragment-spreads in the corresponding definition

Reply:
> variableValues is altered when we are collecting field for said fragment-spread and subsequent nested fragment-spreads in the corresponding definition

I don't think that's what is wanted.

```
query A($x: Int = 3) {
  ...B
}

fragment B($x: String) on Query {
  echo(value: $x)
  ...C
}

fragment C on Query {
  howMany(count: $x)
}
```

`howMany(count: $x)` ought to resolve to `howMany(count: 3)` not `howMany(count: null)`

Reply:
That is exactly what is happening, the tests are all passing. When a variable isn't redefined it implicitly transfers to the child-scope in shadowing

Reply:
Ah sorry I was misunderstanding which scope `variableValues` implied.